panic (index out of range) on incorrect parameter

Using Go and the cockroachdb/pg driver, if you use $0 as a parameter it crashes the database server.

e.g.

```
db.Exec(`INSERT INTO person(name, email) VALUES($0, $1)`, name, email)
```

```
panic: runtime error: index out of range

goroutine 37179 [running]:
panic(0x1636820, 0xc82000e090)
        /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleParse(0xc820318000, 0xc820318028, 0x0, 0x0)
        /usr/local/gopath/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:349 +0xf65
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc820318000, 0x0, 0x0, 0x0)
        /usr/local/gopath/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:251 +0xf1b
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc820333a58, 0x7ff1e708d578, 0xc82019e580, 0x0, 0x0)
        /usr/local/gopath/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:161 +0xb0b
github.com/cockroachdb/cockroach/server.(*Server).Start.func4.1(0x7ff1e708d578, 0xc82019e580)
        /usr/local/gopath/src/github.com/cockroachdb/cockroach/server/server.go:292 +0x4b
github.com/cockroachdb/cockroach/util.ServeHandler.func3.1(0xc82014c080, 0x7ff1e708d578, 0xc82019e580, 0xc8202b6020)
        /usr/local/gopath/src/github.com/cockroachdb/cockroach/util/net.go:123 +0x5a
created by github.com/cockroachdb/cockroach/util.ServeHandler.func3
        /usr/local/gopath/src/github.com/cockroachdb/cockroach/util/net.go:125 +0x31f

```

--- a/sql/parser/type_check.go
+++ b/sql/parser/type_check.go
@@ -94,7 +94,7 @@ func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
 			return nil, err
 		}
 		if _, ok := dummyCond.(DValArg); ok {
-			return nil, fmt.Errorf("cannot infer type for parameter %s", dummyCond)
+			return nil, fmt.Errorf("could not determine data type of parameter %s", dummyCond)
 		}
 		condArgType = dummyCond
 	}
@@ -107,7 +107,7 @@ func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
 			return nil, err
 		}
 		if _, ok := dummyVal.(DValArg); ok {
-			return nil, fmt.Errorf("cannot infer type for parameter %s", dummyVal)
+			return nil, fmt.Errorf("could not determine data type of parameter %s", dummyVal)
 		}
 	}
 
@@ -132,7 +132,7 @@ func (expr *CaseExpr) TypeCheck(args MapArgs) (Datum, error) {
 			return nil, err
 		}
 		if _, ok := nextDummyVal.(DValArg); ok && dummyVal == DNull {
-			return nil, fmt.Errorf("cannot infer type for parameter %s", nextDummyVal)
+			return nil, fmt.Errorf("could not determine data type of parameter %s", nextDummyVal)
 		}
 		if set, err := args.SetInferredType(nextDummyVal, dummyVal); err != nil {
 			return nil, err

--- a/sql/pgwire/v3.go
+++ b/sql/pgwire/v3.go
@@ -329,7 +329,7 @@ func (c *v3Conn) handleParse(buf *readBuffer) error {
 	}
 	pq := preparedStatement{
 		query:       query,
-		inTypes:     make([]oid.Oid, len(args)),
+		inTypes:     make([]oid.Oid, 0, len(args)),
 		portalNames: make(map[string]struct{}),
 		columns:     cols,
 	}
@@ -337,12 +337,16 @@ func (c *v3Conn) handleParse(buf *readBuffer) error {
 	for k, v := range args {
 		i, err := strconv.Atoi(k)
 		if err != nil {
-			return c.sendError(fmt.Sprintf("non-integer parameter name: %s", k))
+			return c.sendError(fmt.Sprintf("non-integer parameter: %s", k))
 		}
 		// ValArgs are 1-indexed, pq.inTypes are 0-indexed.
 		i--
-		if len(pq.inTypes) <= i {
-			return c.sendError("internal error: leftover valargs")
+		if i < 0 {
+			return c.sendError(fmt.Sprintf("there is no parameter $%s", k))
+		}
+		// Grow pq.inTypes to be at least as large as i.
+		for j := len(pq.inTypes); j <= i; j++ {
+			pq.inTypes = append(pq.inTypes, 0)
 		}
 		// OID to Datum is not a 1-1 mapping (for example, int4 and int8 both map
 		// to DummyInt), so we need to maintain the types sent by the client.
@@ -354,6 +358,11 @@ func (c *v3Conn) handleParse(buf *readBuffer) error {
 			return c.sendError(fmt.Sprintf("unknown datum type: %s", v.Type()))
 		}
 		pq.inTypes[i] = id
+	}
+	for i := range pq.inTypes {
+		if pq.inTypes[i] == 0 {
+			return c.sendError(fmt.Sprintf("could not determine data type of parameter $%d", i+1))
+		}
 	}
 	c.preparedStatements[name] = pq
 	c.writeBuf.initMsg(serverMsgParseComplete)

--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -217,13 +217,15 @@ func TestPGPrepareFail(t *testing.T) {
 		"SELECT $1":                                     "pq: could not determine data type of parameter $1",
 		"SELECT $1 + $1":                                "pq: unsupported binary operator: <parameter> + <parameter>",
 		"SELECT now() + $1":                             "pq: unsupported binary operator: <timestamp> + <parameter>",
-		"SELECT CASE $1 WHEN 1 THEN 1 END":              "pq: cannot infer type for parameter $1",
-		"SELECT CASE WHEN TRUE THEN $1 END":             "pq: cannot infer type for parameter $1",
-		"SELECT CASE WHEN TRUE THEN $1 ELSE $2 END":     "pq: cannot infer type for parameter $2",
-		"SELECT CASE WHEN TRUE THEN 1 ELSE $1 END":      "pq: cannot infer type for parameter $1",
-		"UPDATE d.t SET d = CASE WHEN TRUE THEN $1 END": "pq: cannot infer type for parameter $1",
+		"SELECT CASE $1 WHEN 1 THEN 1 END":              "pq: could not determine data type of parameter $1",
+		"SELECT CASE WHEN TRUE THEN $1 END":             "pq: could not determine data type of parameter $1",
+		"SELECT CASE WHEN TRUE THEN $1 ELSE $2 END":     "pq: could not determine data type of parameter $2",
+		"SELECT CASE WHEN TRUE THEN 1 ELSE $1 END":      "pq: could not determine data type of parameter $1",
+		"UPDATE d.t SET d = CASE WHEN TRUE THEN $1 END": "pq: could not determine data type of parameter $1",
 		"CREATE TABLE $1 (id INT)":                      "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
 		"UPDATE d.t SET s = i + $1":                     "pq: value type int doesn't match type STRING of column \"s\"",
+		"SELECT $0 > 0":                                 "pq: there is no parameter $0",
+		"SELECT $2 > 0":                                 "pq: could not determine data type of parameter $1",
 	}
 
 	if _, err := db.Exec(`CREATE DATABASE d; CREATE TABLE d.t (i INT, s STRING, d INT)`); err != nil {

